packages/next/build/webpack/plugins/flight-client-entry-plugin.ts

```diff
@@ -55,6 +62,47 @@ export class FlightClientEntryPlugin {
           (webpack as any).dependencies.ModuleDependency,
           new (webpack as any).dependencies.NullDependency.Template()
         )
+
+        normalModuleFactory.hooks.resolve.tap(
+          PLUGIN_NAME,
+          function (resolveData: any) {
+            if (
+              resolveData.contextInfo.issuerLayer === WEBPACK_LAYERS.server &&
+              // TODO-APP: Handle edge-server here correctly.
+              resolveData.contextInfo.compiler === 'server'
+            ) {
+              if (
+                [
+                  PAGES_DIR_ALIAS,
+                  ROOT_DIR_ALIAS,
+                  DOT_NEXT_ALIAS,
+                  APP_DIR_ALIAS,
+                  RSC_MOD_REF_PROXY_ALIAS,
+                ].some((alias) => resolveData.request.startsWith(alias))
+              ) {
+                return
+              }
+
+              resolveData.resolveOptions = {
+                ...resolveData.resolveOptions,
+                conditionNames: ['react-server'],
+                alias: process.env.__NEXT_REACT_CHANNEL
+                  ? {
+                      react: `react-${process.env.__NEXT_REACT_CHANNEL}`,
+                      'react/package.json': `react-${process.env.__NEXT_REACT_CHANNEL}/package.json`,
+                      'react/jsx-runtime': `react-${process.env.__NEXT_REACT_CHANNEL}/jsx-runtime`,
+                      'react/jsx-dev-runtime': `react-${process.env.__NEXT_REACT_CHANNEL}/jsx-dev-runtime`,
+                      'react-dom': `react-dom-${process.env.__NEXT_REACT_CHANNEL}`,
+                      'react-dom/package.json': `react-dom-${process.env.__NEXT_REACT_CHANNEL}/package.json`,
+                      'react-dom/server': `react-dom-${process.env.__NEXT_REACT_CHANNEL}/server`,
+                      'react-dom/server.browser': `react-dom-${process.env.__NEXT_REACT_CHANNEL}/server.browser`,
+                      'react-dom/client': `react-dom-${process.env.__NEXT_REACT_CHANNEL}/client`,
+                    }
+                  : false,
+              }
+            }
+          }
+        )
```

Comment:
I don't think that's the best way. Have you tried adding a `module.rules` rule which changes the resolve alias for the server layer:

```
module.rules: [
  {
    issuerLayer: WEBPACK_LAYERS.server,
    resolve: {
      alias: { react: ..., ... }
    }
  },
  ...
]
```

Reply:
That's a better way that I didn't consider - is it possible to specify conditionNames in the module rule (just alias isn't enough here) ? I don't see it in the docs though

Reply:
`conditionNames` is part of resolve option, should be able to leverage in `resolve.conditionNames` 🤔